api/src/main/java/com/epam/pipeline/manager/cluster/performancemonitoring/ResourceMonitoringManager.java

```diff
@@ -105,40 +104,92 @@ public void init() {
         instanceTypesObservable.subscribe(instanceTypes -> instanceTypeMap = instanceTypes.stream()
                 .collect(Collectors.toMap(InstanceType::getName, t -> t, (t1, t2) -> t1)));
 
-        scheduleFixedDelaySecured(this::monitorResourceUsage, SystemPreferences.SYSTEM_RESOURCE_MONITORING_PERIOD,
+        scheduleFixedDelaySecured(this::monitorResourceUsage, SYSTEM_RESOURCE_MONITORING_PERIOD,
                 "Resource Usage Monitoring");
     }
 
     @Scheduled(cron = "0 0 0 ? * *")
     public void removeOldIndices() {
         monitoringDao.deleteIndices(preferenceManager.getPreference(
-            SystemPreferences.SYSTEM_RESOURCE_MONITORING_STATS_RETENTION_PERIOD));
+            SYSTEM_RESOURCE_MONITORING_STATS_RETENTION_PERIOD));
     }
 
     public void monitorResourceUsage() {
-        int idleTimeout = preferenceManager.getPreference(SYSTEM_MAX_IDLE_TIMEOUT_MINUTES);
         Map<String, PipelineRun> running = pipelineRunManager.loadRunningPipelineRuns().stream()
-            .filter(r -> DateUtils.nowUTC().isAfter(r.getProlongedAtTime().plus(idleTimeout, ChronoUnit.MINUTES)))
-            .collect(Collectors.toMap(PipelineRun::getPodId, r -> r));
+                .collect(Collectors.toMap(PipelineRun::getPodId, r -> r));
 
-        log.debug("Checking cpu stats for pipelines: " + running.keySet().stream()
-            .collect(Collectors.joining(", ")));
+        processIdleRuns(running);
+        processOverloadedRuns(running);
+    }
+
+    private void processOverloadedRuns(Map<String, PipelineRun> running) {
+        final List<Pair<PipelineRun, Map<String, Double>>> runsToNotify = new ArrayList<>();
+        int timeRange = preferenceManager.getPreference(SYSTEM_MONITORING_METRIC_TIME_RANGE);
+        final Map<String, Double> thresholds = getThresholds();
+        log.debug("Checking memory and disk stats for pipelines: " + String.join(", ", running.keySet()));
+
+        LocalDateTime now = DateUtils.nowUTC();
+        Map<String, Map<String, Double>> metrics = Stream.of(ELKUsageMetric.MEM, ELKUsageMetric.FS)
+                .collect(Collectors.toMap(ELKUsageMetric::getName, metric ->
+                        monitoringDao.loadUsageRateMetrics(metric, running.keySet(),
+                            now.minusMinutes(timeRange), now)));
+
+        running.forEach((pod, run) -> {
```

Comment:
Prefer usage of `map` to create a new collection, instead of mutating an existing one